### ISSUE TITLE
[enh.] Fixed IP forced by User

### ISSUE THREAD

**Issue by PizzaProgram** [SoftEtherVPN/SoftEtherVPN#1797]
This idea is easy to implement, and would have 2 purposes:
- Fixed IP for SecureNAT DHCP 
- Check multi-login attempts (if the configuration was duplicated/**stolen** and two clients try to log in with same username.)

## The Setup:
1. Simply filling the "Force IP" field would activate this feature.
2. _Optional:_ By filling it, it would automatically Check the "Warn on multi login" checkbox.
3. _Optional:_ This setup window would disallow to click [OK] button, if the IP field is not a valid IP _(or would simply delete the field.)_
4. _Optional:_ By pressing the OK button, it could warn the administrator:
- if SecureNAT is not active 
- or the IP is not at the same DHCP subnet 
- or the IP is **in the** DHCP's **range** ! _(Disallow collusion)_
- or this IP is occupied by an other User
![kép](https://user-images.githubusercontent.com/26026089/222695933-5f049fd0-6f5e-47df-8893-0a625fa53893.png)

##The Code:

SecureNAT DHCP would first check if `ForcedIP` is not empty
```
declare UserList
{
	declare User123
	{
		{
		string ForcedIP 10.11.12.13
		bool WarnMultiLogin true
...
```

... and if that's the case it would add that IP to the connected user.
Except if an other user is occupying that IP already -> than:
- **disallow to connect at all !**
- send the client an error message in return: "Only 1 connection is allowed to this user"

So basically it would work the same, as these 3 policies enabled:
![kép](https://user-images.githubusercontent.com/26026089/222716377-c715a892-25a0-41e5-915b-39e3a13dc6e3.png)

This enhancement would be easier + quicker to implement, and could be done **before** fixing the DHCP MAC address reservation issue: #1792

**Comment by PizzaProgram** [comment]
# II. Approach :

Of course it would be much easier to simply add this option directly to the list of Security Policy of the user, (see picture above)
"Enforce DHCP Fixed IP (IPv4)"
"Enforce DHCP Fixed IP (IPv6)"

but in that case the management of the occupied IP list would be more difficult, 
so a "List window" would be needed to see all the pre-set IPs and users.
_(Where the list could be sorted via IP column)_

**Comment by PizzaProgram** [comment]
Later we should add these parameters to the HTML API too. (Both: setting IP + reading those values.)
That would help enterprise-level IT management a lot.

**Comment by PizzaProgram** [comment]
_Instead of rewriting the whole SecureNAT DHCP completely to store leases (as described at #1792 ) ,_
can rather somebody add those few lines of code to make [this II. Approach](https://github.com/SoftEtherVPN/SoftEtherVPN/issues/1797#issuecomment-1453611876) work?  ❓ 
  
- [ ] Add entries to "Security Policy of the Users" : `"Enforce DHCP Fixed IP (IPv4)" | 10.11.12.13`
- [ ] At SecureNAT DHCP: load those values from config file when client is connecting to add that IP
- [ ] Add +2 columns to "Users" list to make it easy to manage

I've searched in the code myself, but could not find where to make those (nearly Copy-Paste) changes myself.  😿 
@chipitsine @davidebeatrici  
_I may underestimate this a bit, but my guess is, that you guy (as C pros) would need ca 5+5+5 minutes top..._

**Comment by shuricksumy** [comment]
Hi,
I looked for the same functionality, and it seems there is some workaround already existing in the code. Sorry, if someone is aware of it, but I decided to share my settings that work well at v5.02.5181.

So the main idea is here in [COMMIT](https://github.com/SoftEtherVPN/SoftEtherVPN/commit/b890c7d81381d6d34a40c77ed86d3e65760565f4?diff=unified&w=1): to use some description for each client formatted as IPv4:xxx.xxx.xxx.xxx. I tried, and this works. Here is an example of my settings:

- Client note (Description): IPv4:10.8.8.39
- Use Virtual NAT is enabled
- Use Virtual DHCP is enabled
- Distributed address: from 10.8.8.100 to 10.8.8.200

It works for L2TP/IPsec and for OpenVPN L3 clients

**Comment by PizzaProgram** [comment]
@shuricksumy Many thanks for sharing this knowledge with us!
May I ask:
 - How did you test it? Did you apply the commit yourself and compiled a binary from it?
 - Is that binary available somewhere?
 - Maybe a docker container?

**Comment by shuricksumy** [comment]
> @shuricksumy Many thanks for sharing this knowledge with us! May I ask:
> 
> * How did you test it? Did you apply the commit yourself and compiled a binary from it?

This code in repo from Oct 12, 2020

> * Is that binary available somewhere?

For Windows, there is some night builder according to manuals: [LINK](https://dev.azure.com/SoftEther-VPN/SoftEther%20VPN/_build/results?buildId=1134&view=artifacts&pathAsName=false&type=publishedArtifacts)

I build binary files for Linux based on the official manuals: [LINK](https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/BUILD_UNIX.md)

> * Maybe a docker container?

Yes, there is an official one [HERE](https://github.com/SoftEtherVPN/SoftetherVPN-docker). Use here NOT stable version for v.5

And [THIS](https://github.com/cenk1cenk2/docker-softether-vpnsrv)  is a good example of how to create Docker and compile the latest version (olso use not stable for v.5), but this uses some additional features like firewall configuration and third-party DHCP server

**Comment by PizzaProgram** [comment]
An interesting test:

- What happens, if I lower the (server VirtualNAT) DHCP lease time ?

## Test result of 40 sec :  
 - the client is reconnecting fine and does get the desired 
 - Reconnection time is set for 15 sec 
 - but it stays disconnected for 30 sec, and connected only for 15 sec

![kép](https://github.com/SoftEtherVPN/SoftEtherVPN/assets/26026089/734f0030-92c7-4378-8b43-be28a254e4f8)

**Comment by PizzaProgram** [comment]
## Test result at 30 sec DHCP lease time:

... Client reconnect = 15 sec
```
 9 sec OFF
11 sec online
31 sec OFF
14 sec online
```
![kép](https://github.com/SoftEtherVPN/SoftEtherVPN/assets/26026089/97999fda-642b-4ebe-af26-48656172d6d9)
